restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/AccountsEndpoint.java

```diff
@@ -597,6 +583,118 @@ protected Response updateAccount(final String identifier, final MultivaluedMap<S
+
+        if (organization != null) {
+            return organization;
+        } else {
+            //Attempt to match organizationId to organizationSid
+            try {
+                organizationSid = new Sid(organizationId);
+                organization = organizationsDao.getOrganization(organizationSid);
+                return organization;
+            } catch (IllegalArgumentException e) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Exception while tried to match organizationId to organizationSid");
+                }
+            }
+        }
+        return organization;
+    }
+
+    protected Response migrateAccountOrganization(final String identifier, final MultivaluedMap<String, String> data,
+                                              final MediaType responseType) {
+
+        //Validation 1 - Only SuperAdmin is allowed to migrate organization for an Account
+        if (!isSuperAdmin()) {
+            throw new InsufficientPermission();
+        }
+
+        Organization organization = getOrganization(data);
+        //Validation 2 - Check if data contains Organization (either SID or domain name)
+        if (organization == null) {
+            return status(PRECONDITION_FAILED).entity("Missing Organization SID or Domain Name").build();
+        }
+
+        Account operatingAccount = getOperatingAccount(identifier);
+
+        //Validation 3 - Operating Account shouldn't be null;
+        if (operatingAccount == null) {
+            return status(NOT_FOUND).build();
+        }
+
+        //Validation 4 - Only direct child of super admin account can be migrated to a new organization
+        if (!isDirectChildOfAccount(userIdentityContext.getEffectiveAccount(), operatingAccount)) {
+            return status(BAD_REQUEST).build();
+        }
+
+        //Validation 5 - Check if Account already in the requested Organization
+        if (operatingAccount.getOrganizationSid().equals(organization.getSid())) {
+            return status(BAD_REQUEST).entity("Account already in the requested Organization").build();
+        }
+
+        //Update Account for the new Organization
+        Account modifiedAccount = operatingAccount.setOrganizationSid(organization.getSid());
+        accountsDao.updateAccount(modifiedAccount);
+
+        if (logger.isDebugEnabled()) {
+            String msg = String.format("Parent Account %s migrated to Organization %s", modifiedAccount.getSid(), organization.getSid());
+            logger.debug(msg);
+        }
+
+        //Update Child accounts and their numbers
+        List<Account> childAccounts = accountsDao.getChildAccounts(operatingAccount.getSid());
+        for (Account child : childAccounts) {
+            if (!child.getOrganizationSid().equals(organization.getSid())) {
+                Account modifiedChildAccount = child.setOrganizationSid(organization.getSid());
+                accountsDao.updateAccount(modifiedChildAccount);
```

Comment:
is this part of the same DB transaction as all the DAO updates in this business case? how do we tackle intermidiate failures? Are we expecting failures in the middle?

Reply:
@jaimecasero all the DAO updates are separate DB transactions and there is no handling for failures in the middle.
What is your suggestion for this?
I would avoid to over-complicate the feature since 99% this API will only be used ONCE to move the accounts to a specific organization and never again.
Maybe we should delegate verification and failure handling to the admin/script that will use the API.